### PR TITLE
alt-server: init at 0.0.5

### PR DESCRIPTION
**Pull request by max-amb**
## Description of changes

https://github.com/NyaMisty/AltServer-Linux

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/misc/alt-server/default.nix [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{ stdenv, fetchurl, lib, pkgs, autoPatchelfHook }:
+
+stdenv.mkDerivation rec {
```
Comment (drupol) [inline pkgs/tools/misc/alt-server/default.nix:3]: I think we should use the `finalAttrs` pattern here.

The `finalAttrs` pattern will let you remove the `rec` keyword ([see its implementation in Nix](https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libexpr/nixexpr.cc#L121)).

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

and then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Feel free to reach out if you need some assistance.

File: pkgs/tools/misc/alt-server/default.nix [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  pname = "AltServer-Linux";
+  version = "0.0.5";
+
+  src = fetchurl {
+    url = "https://github.com/NyaMisty/AltServer-Linux/releases/download/v${version}/AltServer-x86_64";
+    sha256 = "sha256-C+fDrcaewRd6FQMrO443xdDk/vtHycQ5zWLCOLPqF/s=";
```
Comment (drupol) [inline pkgs/tools/misc/alt-server/default.nix:9]: I think you should use `hash` in place of `sha256`.

This is optional since it could be automatized treewide at some point.

The idea is to avoid using `sha256` as attribute and use `hash` everywhere.

Since there could be different hashing algorithms in use, this make sense.

File: pkgs/tools/misc/alt-server/default.nix [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  };
+
+  dontUnpack = true;
+
+  installPhase = ''
```
Comment (drupol) [inline pkgs/tools/misc/alt-server/default.nix:14]: Missing phase hooks, see https://github.com/jtojnar/nixpkgs-hammering/blob/main/explanations/missing-phase-hooks.md

File: pkgs/tools/misc/alt-server/default.nix [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    mkdir -p $out/bin
+    cp $src $out/bin/alt-server
+    chmod u+rx $out/bin/alt-server
+  '';
+
+  BuildInputs = [
```
Comment (drupol) [inline pkgs/tools/misc/alt-server/default.nix:20]: Mind the case.

it's `buildinputs`, not `BuildInputs`.

File: pkgs/tools/misc/alt-server/default.nix [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    pkgs.avahi-compat.dev
+  ];
+
+  meta = with lib; {
```
Comment (drupol) [inline pkgs/tools/misc/alt-server/default.nix:24]: Consider adding yourself maintainer.

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -272,6 +272,8 @@ with pkgs;
 
   ast-grep = callPackage ../development/tools/misc/ast-grep { };
 
+  alt-server = callPackage ../tools/misc/alt-server { };
```
Comment (drupol) [inline pkgs/top-level/all-packages.nix:275]: The letter `l` comes before the letter `s`. Please update.

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your first contribution to the Nix project.

I've left some comments, left me know if this is clear enough.

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -272,6 +274,7 @@ with pkgs;
 
   ast-grep = callPackage ../development/tools/misc/ast-grep { };
 
+
```
Comment (drupol) [inline pkgs/top-level/all-packages.nix:277]: This is an unrelated change, please revert.

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix (hunk 7 above)
Comment (max-amb) [reply pkgs/top-level/all-packages.nix:277]: Oh i believed that that was what you were asking for in the previous changes, sorry!

File: pkgs/tools/misc/alt-server/default.nix (hunk 3 above)
Comment (max-amb) [reply pkgs/tools/misc/alt-server/default.nix:18]: Could you check that I have implemented this correctly in [this](https://github.com/NixOS/nixpkgs/pull/246161/commits/ac5f658cb2326880f473bcf6959fd043f58fee1e) commit?

File: pkgs/top-level/all-packages.nix (hunk 6 above)
Comment (max-amb) [reply pkgs/top-level/all-packages.nix:275]: I believed this was fixed in [this](https://github.com/NixOS/nixpkgs/pull/246161/commits/70972f26e0ecdd586bc94fdd01d7167a93431bb7) commit, but I seem to have been mistaken, could you explain what you mean?

File: pkgs/top-level/all-packages.nix (hunk 7 above)
Comment (max-amb) [reply pkgs/top-level/all-packages.nix]: Never mind, my mistake

File: pkgs/tools/misc/alt-server/default.nix [hunk 8]
```diff
@@ -0,0 +1,38 @@
+{ stdenv, fetchurl, lib, pkgs, autoPatchelfHook }:
```
Comment (drupol) [inline pkgs/tools/misc/alt-server/default.nix:1]:
```suggestion
{ stdenv
, fetchurl
, lib
, avahi-compat
, autoPatchelfHook
}:

```

File: pkgs/tools/misc/alt-server/default.nix [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+  pname = "AltServer-Linux";
+  version = "0.0.5";
+
+  src = fetchurl {
+    url = "https://github.com/NyaMisty/AltServer-Linux/releases/download/v${finalAttrs.version}/AltServer-x86_64";
+    hash = "sha256-C+fDrcaewRd6FQMrO443xdDk/vtHycQ5zWLCOLPqF/s=";
+  };
+
+  buildInputs = [
+    pkgs.avahi-compat.dev
```
Comment (drupol) [inline pkgs/tools/misc/alt-server/default.nix:13]:
```suggestion
    avahi-compat
```

**Review by drupol — CHANGES_REQUESTED** [review]
Once last change and it should be ok.

Don't forget that this PR should only contains 2 commits:

- 1 for adding you as maintainer: `maintainers: add max-amb`
- 1 for the rest: `alt-server: init at 0.0.5`

**Comment by drupol** [comment]
@yu-re-ka What happened here?

**Comment by max-amb** [comment]
Sorry I was just resetting the commits to only have two and it seems to have coincided badly with the merge, oops.

**Comment by drupol** [comment]
@max-amb Sorry for this, feel free to recreate a PR and ping me in it.

**Comment by max-amb** [comment]
OK will do!
